cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/PartitionConsumerState.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright © 2015 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.api.dataset.lib;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Sets;
+import com.google.common.primitives.Longs;
+
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.Set;
+
+/**
+ * Contains the state necessary to keep track of which partitions are processed and which partitions would need to be
+ * processed as they are created.
+ */
+public class PartitionConsumerState {
+  // useful on initial query of partitions
+  public static final PartitionConsumerState FROM_BEGINNING =
+    new PartitionConsumerState(0, Collections.<Long>emptySet());
+
+  // Read pointer of the transaction from the previous query of partitions. This is used to scan for new partitions
+  // created since then.
+  private final long startVersion;
+  // The list of in progress transactions from the previous query of partitions that are smaller than the read pointer.
+  // We do not need to include the in-progress transaction Ids that are larger than the read pointer because those will
+  // be picked up in the next scan anyways, since we will start the scan from the read pointer.
+  // Keeping track of these in-progress transactions is necessary because these might be creations of partitions that
+  // fall before the startVersion.
+  private final Set<Long> versionsToCheck;
+
+  public PartitionConsumerState(long startVersion, Set<Long> versionsToCheck) {
+    this.startVersion = startVersion;
+    this.versionsToCheck = versionsToCheck;
+  }
+
+  public long getStartVersion() {
+    return startVersion;
+  }
+
+  public Set<Long> getVersionsToCheck() {
+    return versionsToCheck;
+  }
+
+  public static PartitionConsumerState fromBytes(byte[] bytes) {
+    Preconditions.checkArgument(bytes.length % Longs.BYTES == 0, "bytes does not have length divisible by Longs.BYTES");
+    ByteBuffer bb = ByteBuffer.wrap(bytes);
+    long startVersion = bb.getLong();
+    Set<Long> versionsToCheck = Sets.newHashSet();
+    while (bb.hasRemaining()) {
+      versionsToCheck.add(bb.getLong());
+    }
+    return new PartitionConsumerState(startVersion, versionsToCheck);
+  }
+
+  public byte[] toBytes() {
+    int numLongs = 1 + versionsToCheck.size();
```

Comment:
Can we add an additional byte at the beginning for the version of the serialization format? That can be 0 for now, but if we ever change this format, then we can bump it to 1, and have smart logic to deserialize the current or previous versions.